testing/upgradetest/versions.go

```diff
@@ -36,33 +41,102 @@ var (
 	Version_8_13_0 = version.NewParsedSemVer(8, 13, 0, "", "")
 )
 
-// GetUpgradableVersions returns the version that the upgradeToVersion can upgrade from.
-func GetUpgradableVersions(ctx context.Context, upgradeToVersion string, currentMajorVersions int, previousMajorVersions int) ([]*version.ParsedSemVer, error) {
-	aac := tools.NewArtifactAPIClient()
-	vList, err := aac.GetVersions(ctx)
+// VersionRequirements is to set requirements for upgradable versions while fetching them.
+//
+// Keep in mind that requirements can overlap. For example 2 previous minors might already
+// cover 2 current majors, so the results would be combined.
+type VersionRequirements struct {
+	UpgradeToVersion string
+	CurrentMajors    int
+	PreviousMajors   int
+	PreviousMinors   int
+	RecentSnapshots  int
+}
+
+const AgentVersionsFilename = ".agent-versions.json"
+
+type AgentVersions struct {
+	// TestVersions contains semver-compliant versions of the agent to run integration tests against.
+	TestVersions []string `json:"testVersions"`
+}
+
+var (
+	agentVersions       *AgentVersions
+	ensureAgentVersions = sync.OnceFunc(setAgentVersions)
+)
+
+func setAgentVersions() {
+	v, err := getAgentVersions()
 	if err != nil {
-		return nil, fmt.Errorf("error retrieving versions from Artifact API: %w", err)
+		panic(err)
```

Comment:
[Question]
Why panic instead of returning an error?

I believe it would be better to return an error with some added context.

Reply:
The answer is on the line 65. This function is used in the `OnceFunc` that is assigned on the package import. I don't know of a better way how to handle errors in this case.

Reply:
Addressed in https://github.com/elastic/elastic-agent/pull/4331/commits/e5d10edb87ca03f7a2e8135305bdd3d59e958f2f